webrtc-stats.html

```diff
@@ -612,8 +612,54 @@ <h3>
         </p>
         <p>
-          RTP monitored objects are not [= deleted =].
+          RTP monitored objects are deleted when the RTP stream is reconfigured
+          such that the {{RTCRtpStreamStats/ssrc}} or
+          {{RTCRtpStreamStats/codecId}} changes or if the
+          {{RTCRtpTransceiver}}'s {{RTCRtpTransceiver/currentDirection}} becomes
+          <code>"stopped"</code>. The monitored object is <i>not</i> deleted if
+          the transceiver is made <code>"inactive"</code> or if the encoding
+          parameters' {{RTCRtpEncodingParameters/active}} attribute is set to
+          <code>false</code>. When an RTP stream is reconfigured such that the
+          monitored object is deleted but the RTP stream continues to be used, a
+          new RTP monitored object and {{RTCStats/id}} will be created for the
+          new SSRC or codec.
         </p>
+        <p>
+          For a given RTP stats object its total counters must always increase,
+          but due to changes in SSRC or codec, an RTP stats object can be
+          deleted and replaced by another RTP stats object whose counters have
+          been reset. The caller should be aware of this when aggregating packet
+          counters accross multiple RTP stats objects.
+        </p>
+        <pre class="example highlight">
+          An RTCRtpSender is sending two layers of simulcast on SSRC=111 and
+          SSRC=222. Two "outbound-rtp" stats objects are observed with ssrc=111
+          and ssrc=222. Both objects' packet counters are increasing.
+
+          One layer is inactivated using RTCRtpSender.setParameters(). While
+          this pauses one of the layers (its packet counter freezes), the RTP
+          monitored objects are not deleted. The RTCRtpTransceiver is negotiated
+          as "inactive" and the RTP monitored objects are still not deleted.
+          When the RTCRtpTransceiver becomes "sendonly" again, the same
+          "outbound-rtp" objects continue to be used.
+
+          Later, RTCRtpTransceiver.stop() is called. The "outbound-rtp" objects
+          still exist but their packet counters have frozen. Renegotiation
+          happens and the transceiver.currentDirection becomes "stopped", now
+          both "outbound-rtp" objects have been deleted.
+        </pre>
+        <pre class="example highlight">
+          An RTCRtpReceiver is receiving VP8 on SSRC=111 with PT=96. We observe
+          one "inbound-rtp" with ssrc=111 and a codecId reference to an
+          RTCCodecStats with payloadType=96.
+
+          Without renegotiating, we start receiving VP9 with PT=100 instead, on
+          the same SSRC. The existing "inbound-rtp" is deleted and a new
+          "inbound-rtp" stats object is created. The new "inbound-rtp" has the
+          same ssrc value, but a different stats ID and its packet counter
+          starts from zero. The new object's codecId references a different
+          RTCCodecStats with payloadType=100.
```

Comment:
I know you got this question during the meeting as well, but I didn't understand the answer, and we didn't have a lot of time to go deeper. What's the use case rationale to reset counters on codec change?

I say "reset counters", because that's what apps will experience here. Counters will go backwards. That's because the stats model relies on cross-referencing associations, which in this case is done using the SSRC, not the stats-id, right?

This means apps may not even notice a change in the stats-id. Asking apps to differentiate objects by stats-id uniqueness seems new. These are stats for an ssrc I thought,  not an ssrc+codec pair. Am I wrong?

Reply:
> These are stats for an ssrc I thought, not an ssrc+codec pair. Am I wrong?

My reasoning for resetting (i.e. delete + create a replacing stats object) on codec change mainly had to do with "what if a simulcast layer goes away".

The simple case is if a layer is dropped via renegotiation because in this case the layer deletion is explicit in the SDP: if layer A has 10 packets and layer B has 10 packets, that's 20. But if layer B is permanently deleted and only layer A remains, this is now only 10 packets.

I admit **you don't need to** reset layer A here, the app will see that layer B went away so it is already aware that a reconfiguration happened, and should not be surprised that "sum of packets = 10" instead of 20. I mainly considered the reset of stats ID a **good signal** that reconfiguration happened, but we can remove this part of the PR if you prefer? WDYT? No strong opinion here.

In either case, if someone ends up using that SSRC again after this fact, that should be a new stats object, because otherwise the implementation would need to save history indefinitely (memory leak). So I don't consider RTP stats to mean "one SSRC", rather it is "one instance of an SSRC being used".

The other reason for resetting on codec change, and not just layer being dropped, was "what if we switch from simulcast to SVC?". I'm not sure what happens here...
- Let's say you negotiate SSRC 1 and SSRC 2 and you use both. Two outbound-rtp appears.
- Now let's say you change to sending SVC, meaning you only make use of SSRC 1 and not SSRC 2. You're no longer configured to send SSRC 2, but AFAIK the SDP may still reflect simulcast. So does the RTP stream for SSRC 2 exist or not?

My thinking was that deleting on codec change seemed like a good "catch all" to ensure the implementation never has to save history of old SSRCs, but maybe this case is already covered by "reducing layers".

Reply:
Should we only delete on SSRC changing, layer dropping or stop(), and not on codec change?

Reply:
Updated the PR not to delete on codec change, only on SSRC going "away"